openmc/dagmc.py

```diff
@@ -0,0 +1,593 @@
+
+    def _n_geom_elements(self, geom_type):
+        """
+        Helper function for retrieving the number geometric entities in a DAGMC
+        file
+
+        Parameters
+        ----------
+        geom_type : str
+            The type of geometric entity to count. One of {'Volume', 'Surface'}. Returns
+            the runtime number of voumes in the DAGMC model (includes implicit complement).
+
+        Returns
+        -------
+        int
+            Number of geometry elements of the specified type
+        """
+        cv.check_value('geometry type', geom_type, ('volume', 'surface'))
+
+        def decode_str_tag(tag_val):
+            return tag_val.tobytes().decode().replace('\x00', '')
+
+        dagmc_filepath = Path(self.filename).resolve()
+        with h5py.File(dagmc_filepath) as dagmc_file:
+            category_data = dagmc_file['tstt/tags/CATEGORY/values']
+            category_strs = map(decode_str_tag, category_data)
+            n = sum([v == geom_type.capitalize() for v in category_strs])
+
+            # check for presence of an implicit complement in the file and
+            # increment the number of cells if it doesn't exist
+            if geom_type == 'volume':
+                name_data = dagmc_file['tstt/tags/NAME/values']
+                name_strs = map(decode_str_tag, name_data)
+                if not sum(['impl_complement' in n for n in name_strs]):
+                    n += 1
+        return n
+
+    @property
+    def n_cells(self):
+        return self._n_geom_elements('volume')
+
+    @property
+    def n_surfaces(self):
+        return self._n_geom_elements('surface')
+
+    def create_xml_subelement(self, xml_element, memo=None):
+        if memo is None:
+            memo = set()
+
+        if self in memo:
+            return
+
+        memo.add(self)
+
+        # Ensure that the material overrides are up-t-date
+        self.build_overide_mat_from_cells()
+
+        # Set xml element values
+        dagmc_element = ET.Element('dagmc_universe')
+        dagmc_element.set('id', str(self.id))
+
+        if self.auto_geom_ids:
+            dagmc_element.set('auto_geom_ids', 'true')
+        if self.auto_mat_ids:
+            dagmc_element.set('auto_mat_ids', 'true')
+        dagmc_element.set('filename', str(self.filename))
+        if self.material_overrides:
+            mat_element = ET.Element('material_overrides')
+            for key in self.material_overrides:
+                mat_element.set(key, ' '.join(
+                    t for t in self.material_overrides[key]))
+            dagmc_element.append(mat_element)
+        xml_element.append(dagmc_element)
+
+    def build_overide_mat_from_cells(self):
+        """
+        Builds the material override dictionary for cells with multiple instances.
+
+        Returns:
+            None
+        """
+        for cell in self.cells.values():
+            if cell.n_instances > 1 and isinstance(cell.fill, Iterable):
```

Comment:
Is there a reason this is restriced to cells with more than one instance?

Reply:
your comment made me thinking :)

You are right, if `cell.fill` is an `Iterable` then `cell.n_instances` has to have more than one instance.


I though about the eventuallity to remove restrictions on `Iterable`,(and `cell.n_instance`), but as all DAGMCcells, when sync get their `fill` property that would me filling the override materials dict with all the fill of all DAGMCcells overridden or not.